### PR TITLE
Speedreader

### PR DESCRIPTION
**Pull request by boocmp**
Resolves https://github.com/brave/brave-browser/issues/18242,
Resolves https://github.com/brave/brave-browser/issues/18234,
Resolves https://github.com/brave/brave-browser/issues/18243

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/speedreader/speedreader_service.cc [hunk 1]
```diff
@@ -87,12 +87,12 @@ void SpeedreaderService::RegisterProfilePrefs(PrefRegistrySimple* registry) {
 }
 
 void SpeedreaderService::ToggleSpeedreader() {
-  const bool enabled = prefs_->GetBoolean(kSpeedreaderPrefEnabled);
-  prefs_->SetBoolean(kSpeedreaderPrefEnabled, !enabled);
-  if (!enabled)
+  const bool enabled = !prefs_->GetBoolean(kSpeedreaderPrefEnabled);
```
[the remark this answers is not in the document]
Comment (boocmp) [reply components/speedreader/speedreader_service.cc]: Renamed to 'toggled_value'

File: browser/speedreader/speedreader_tab_helper.cc [hunk 2]
```diff
@@ -193,24 +195,123 @@ void SpeedreaderTabHelper::ShowReaderModeBubble() {
...
+  }
+}
+
+void SpeedreaderTabHelper::ReloadContents() {
+  web_contents()->GetController().Reload(content::ReloadType::NORMAL, false);
+}
+
+void SpeedreaderTabHelper::ProcessNavigation(
+    content::NavigationHandle* navigation_handle) {
+  if (!navigation_handle->IsInMainFrame() ||
```
[the remark this answers is not in the document]
Comment (boocmp) [reply browser/speedreader/speedreader_tab_helper.cc:233]: Done

File: components/speedreader/speedreader_service.h [hunk 3]
```diff
@@ -32,7 +31,7 @@ class SpeedreaderService : public KeyedService {
   SpeedreaderService& operator=(const SpeedreaderService&) = delete;
 
  private:
-  PrefService* prefs_ = nullptr;
+  raw_ptr<PrefService> prefs_ = nullptr;
```
[the remark this answers is not in the document]
Comment (goodov) [reply components/speedreader/speedreader_service.h:34]: I don't think we should drop `= nullptr` from `raw_ptrs`.
From [base/memory/raw_ptr.md](https://github.com/chromium/chromium/blob/main/base/memory/raw_ptr.md):
> You should continue to explicitly initialize raw_ptr members to ensure consistent behavior on platforms where USE_BACKUP_REF_PTR is disabled.

From [base/allocator/allocator.gni](https://github.com/chromium/chromium/blob/b2cd954ad38ba16cf5fecf083a2b5399fa15bcdd/base/allocator/allocator.gni#L89):
```
_is_brp_supported = (is_win || is_android) && use_allocator == "partition"

declare_args() {
  # Set use_backup_ref_ptr true to use BackupRefPtr (BRP) as the implementation
  # of raw_ptr<T>, and enable PartitionAlloc support for it.
  use_backup_ref_ptr = _is_brp_supported
}
```

[the remark this answers is not in the document]
Comment (supermassive) [reply components/speedreader/speedreader_service.h:34]: Ok, makes sense

**Review by supermassive — APPROVED** [review]

File: components/de_amp/browser/de_amp_url_loader.h [hunk 4]
```diff
@@ -51,6 +51,7 @@ class DeAmpURLLoader : public body_sniffer::BodySnifferURLLoader {
   void ForwardBodyToClient();
 
   base::WeakPtr<DeAmpThrottle> de_amp_throttle_;
+  base::WeakPtrFactory<DeAmpURLLoader> weak_factory_{this};
```
Comment (iefremov) [inline components/de_amp/browser/de_amp_url_loader.h:54]: not used?

File: components/speedreader/speedreader_util.h [hunk 5]
```diff
@@ -58,6 +58,12 @@ enum class DistillState {
 // Page is in reader mode or speedreader mode.
 bool PageStateIsDistilled(DistillState state);
 
+// Page can't be distilled.
+bool PageDoesntSupportDistillation(DistillState state);
+
+// Page can be destilled.
```
Comment (iefremov) [inline components/speedreader/speedreader_util.h:64]: typo

File: components/speedreader/speedreader_util.h [hunk 6]
```diff
@@ -58,6 +58,12 @@ enum class DistillState {
 // Page is in reader mode or speedreader mode.
 bool PageStateIsDistilled(DistillState state);
 
+// Page can't be distilled.
+bool PageDoesntSupportDistillation(DistillState state);
```
Comment (iefremov) [inline components/speedreader/speedreader_util.h:62]: it's not really clear what method of these two to use and when. It sounds like `! PageSupportsDistillation()` is the same as `PageDoesn'tSupportDistillation()` but from the code this is not true. Can we switch to a single method or pick better names?

File: browser/speedreader/speedreader_tab_helper.h [hunk 7]
```diff
@@ -82,23 +91,38 @@ class SpeedreaderTabHelper
...
 
   bool MaybeUpdateCachedState(content::NavigationHandle* handle);
-  void UpdateActiveState(content::NavigationHandle* handle);
+  void UpdateActiveState(const GURL& url);
   void SetNextRequestState(DistillState state);
 
-  // content::WebContentsObserver
+  void ClearCache();
+  void ReloadContents();
+  void ProcessNavigation(content::NavigationHandle* navigation_handle);
```
Comment (iefremov) [inline browser/speedreader/speedreader_tab_helper.h:99]: comments?

File: browser/speedreader/speedreader_tab_helper.h [hunk 8]
```diff
@@ -82,23 +91,38 @@ class SpeedreaderTabHelper
...
   void DidStopLoading() override;
+  void OnVisibilityChanged(content::Visibility visibility) override;
+  void WebContentsDestroyed() override;
 
   // SpeedreaderResultDelegate:
   void OnDistillComplete() override;
 
+  std::unique_ptr<PrefChangeRegistrar> pref_change_registrar_;
+
+  bool is_active_ = false;
```
Comment (iefremov) [inline browser/speedreader/speedreader_tab_helper.h:119]: comment?

Comment (iefremov) [reply browser/speedreader/speedreader_tab_helper.h:119]: is it `is_visible_` ?

**Review by iefremov — APPROVED** [review]

**Pull request merged**